[#7177] feat(core): Add a Caffeine-based implementation for EntityCache

Abyss-lord opened the pull request:
### What changes were proposed in this pull request?

Fix some bugs in EntityCache and add test cases.

- [X] Implement the methods related to the `SupportsEntityStoreCache` and `SupportsRelationEntityCache` interface in `CaffeineEntityCache`.
- [X] Add unit tests for both index and cache.
- [ ] Use JCStress to perform multi-threaded testing on `CaffeineEntityCache` and the `CacheIndex`.

| Category                | Test Case Description                             | Method(s)                           |
| ----------------------- | ------------------------------------------------- | ----------------------------------- |
| **Basic Functionality** | Basic put/get/getIfPresent operations             | `testPutAndGet`, `testGetIfPresent` |
|                         | Same identifier with different entity types       | `testPutSameIdentifierEntities`     |
|                         | Size counting                                     | `testSize`                          |
|                         | Clear the cache                                   | `testClear`                         |
| **Cache Loading**       | Hit from cache, no store access                   | `testGetFromCache`                  |
|                         | Miss from cache, fallback to store                | `testGetFromStore`                  |
|                         | Reload after invalidation                         | `testRemoveThenReload`              |
| **Invalidation**        | Invalidate by METALAKE/CATALOG/SCHEMA/TABLE level | `testInvalidate*` series            |
|                         | Invalidate non-existent entity                    | `testRemoveNonExistentEntity`       |
| **Relation Handling**   | Put/Get/Invalidate entity relations               | Covered in multiple tests           |
|                         | Fallback to `listEntitiesByRelation` from store   | `testGetFromStore`                  |
| **Eviction Policies**   | Expire by time                                    | `testExpireByTime`                  |
|                         | Expire by size                                    | `testExpireBySize`                  |
|                         | Expire by weight                                  | `testExpireByWeight`                |
|                         | Exceed max weight immediately                     | `testExpireByWeightExceedMaxWeight` |
| **Weight Logic**        | Correctness of entity weight calculation          | `testWeightCalculation`             |
| **Error & Boundary**    | Store throws exception                            | `testLoadEntityThrowException`      |
|                         | Store is null                                     | `testNullEntityStore`               |
|                         | Null argument checks                              | `test*WithNull` methods             |
| **Helper/Validation**   | Identifier equality check                         | `testEquals`                        |

### Why are the changes needed?

Fix: #7177

### Does this PR introduce _any_ user-facing change?

no

### How was this patch tested?

local test.

Abyss-lord commented on core/src/main/java/org/apache/gravitino/GravitinoEnv.java:
```diff
@@ -427,6 +439,11 @@ private void initBaseComponents() {
+      this.entityCache = CaffeineEntityCache.getInstance(config, entityStore);
+    } else {
+      this.entityCache = CaffeineEntityCache.getInstance(config, null);
+    }
```
Fix

Abyss-lord commented:
Hi @xunliu @jerryshao , could you please review this PR when you have time? I’d really appreciate your feedback.

jerqi commented on core/src/main/java/org/apache/gravitino/cache/BaseEntityCache.java:
```diff
@@ -153,6 +154,48 @@ protected static <E extends Entity & HasIdentifier> E convertEntity(Entity entit
+      NameIdentifier ident, Entity.EntityType type) throws IOException {
+    Preconditions.checkArgument(
+        entityStore != null,
+        "EntityStore must not be null if you want to get entities from " + "db");
```
`must not` -> `cannot`

jerqi commented on core/src/main/java/org/apache/gravitino/cache/CaffeineEntityCache.java:
```diff
@@ -137,7 +137,10 @@ private CaffeineEntityCache(Config cacheConfig, EntityStore entityStore) {
-      NameIdentifier ident, Entity.EntityType type, SupportsRelationOperations.Type relType)
+      NameIdentifier ident,
+      Entity.EntityType type,
+      SupportsRelationOperations.Type relType,
```
Is it necessary to have this parameter?

jerqi commented on core/src/main/java/org/apache/gravitino/cache/BaseEntityCache.java:
```diff
@@ -153,6 +154,48 @@ protected static <E extends Entity & HasIdentifier> E convertEntity(Entity entit
+   * @param <E> The type of the related entities.
+   * @throws IOException if an error occurs while loading the entities.
+   */
+  protected <E extends Entity & HasIdentifier> List<E> loadEntitiesByRelation(
```
`liseEntitiesByRelation`?

Abyss-lord replied:
Fix, already rename to `listEntitiesByRelation`.

Abyss-lord commented on core/src/main/java/org/apache/gravitino/cache/BaseEntityCache.java:
```diff
@@ -153,6 +154,48 @@ protected static <E extends Entity & HasIdentifier> E convertEntity(Entity entit
+      throws IOException {
+    Preconditions.checkArgument(
+        entityStore != null,
+        "EntityStore must not be null if you want to get entities from " + "db");
```
Fix.

jerryshao commented on core/src/main/java/org/apache/gravitino/GravitinoEnv.java:
```diff
@@ -427,6 +439,10 @@ private void initBaseComponents() {
   private void initGravitinoServerComponents() {
     // Initialize EntityStore
     this.entityStore = EntityStoreFactory.createEntityStore(config);
+    if (config.get(Configs.CACHE_ENABLED)) {
+      this.entityCache = CaffeineEntityCache.getInstance(config, entityStore);
+      // TODO constructs a CachedEntityStore instance with the entityStore and entityCache
+    }
```
BTW, the `entityCache` will be null if the cache is not enabled. From my point, we should not create a `null` object deliberately, this will make it hard for the downstream developer to use this object, they have to check if this cache is null or not, and it is not a good implementation.

If the cache is not enabled, we should directly pass through to the entity store call. So for the users of this cache, they can directly use it without needing to add more check.

Abyss-lord replied:
Thanks @jerryshao for the suggestion, makes sense!

I've updated the implementation to support pluggable cache via Java SPI. Specifically:
- Introduced a `CacheFactory` and `CacheProvider` interface to support dynamic creation of EntityCache implementations;
- Added a new configuration key `gravitino.cache.typeName` to specify the desired cache backend (e.g., "Caffeine");
- When cache is **enabled**, we instantiate the appropriate `EntityCache` via SPI and inject it into CachedEntityStore;
- When cache is **disabled**, we still create a `CachedEntityStore`, but pass null for the cache instance, so all operations fall through to the underlying store transparently.
This way, downstream logic only needs to work with `EntityStore` and doesn't need to care whether caching is enabled or not — the behavior is internally consistent and clean.

jerryshao commented on core/src/main/java/org/apache/gravitino/storage/relational/CachedEntityStore.java:
```diff
@@ -0,0 +1,246 @@
+  public boolean exists(NameIdentifier ident, Entity.EntityType entityType) throws IOException {
+    return cache.withCacheLock(
+        () -> {
+          if (cache.contains(ident, entityType)) {
```
I saw you already have a same lock in cache, and you're using the lock again in here. So is it necessary to have a lock in the cache class, since you will always have the lock here?

Abyss-lord commented:
Hi @jerryshao , I've completed the code updates and would appreciate your review of the PR when you have a moment.

Abyss-lord commented:
Hi @jerryshao , I've completed the code updates and would appreciate your review of the PR when you have a moment.

xunliu commented:
@jerryshao Please help review this PR, Thanks.

Abyss-lord commented:
Hi @jerryshao , I've completed the code updates and would appreciate your review of the PR when you have a moment. Here's a summary of the commits:

1. Rebased the branch on top of the latest main.
2. Introduced a SPI interface for cache pluggability.
3. Added test cases for both `CacheIndex` and `CaffeineEntityCache`.

jerryshao commented on core/src/main/java/org/apache/gravitino/Configs.java:
```diff
@@ -389,4 +389,12 @@ private Configs() {}
+          .doc("Name of the cache provider to use.")
+          .version(ConfigConstants.VERSION_1_0_0)
+          .stringConf()
+          .createWithDefault("Caffeine");
```
Please change to "caffeine", do not capitalize, which will easily lead to mis-configuraition.

jerryshao commented on core/src/main/java/org/apache/gravitino/cache/provider/CacheFactory.java:
```diff
@@ -0,0 +1,60 @@
+    }
+    return provider;
+  }
+}
```
I think it is not necessary to use SPI to load the different cache interfaces, we can use class name and reflection to instantiate the class.

jerryshao replied:
You can refer to the implementation like entity storage factory.

Abyss-lord replied:
Fix.

jerryshao commented on core/src/main/java/org/apache/gravitino/cache/provider/CacheProvider.java:
```diff
@@ -0,0 +1,45 @@
+   * @return A cache instance initialized according to the configuration.
+   */
+  EntityCache getCache(Config config);
+}
```
We don't need this interface, it is only used for SPI, but I think using SPI is not necessary. SPI will load all the implementations, which is not what we wanted.

Abyss-lord replied:
Fix. I remove this interface and avoid using SPI. Instead, i directly instantiate the specific implementation via reflection based on configuration.

Abyss-lord commented:
Hi @jerryshao , I've completed the code updates and would appreciate your review of the PR when you have a moment. Here's a summary of the commits:
1. Remove SPI interface.
